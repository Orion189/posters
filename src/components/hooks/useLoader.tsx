import { useState, ReactNode, FC, Dispatch, SetStateAction, createContext, useContext } from 'react';

type LoaderContextProps = {
    isLoaderVisible: boolean;
    setLoaderVisibility: Dispatch<SetStateAction<boolean>>;
};

type LoaderProviderProps = {
    children?: ReactNode;
};

const LoaderContext = createContext<LoaderContextProps | undefined>(undefined);

export const LoaderProvider: FC<LoaderProviderProps> = ({ children }) => {
    const [isLoaderVisible, setLoaderVisibility] = useState(false);

    return <LoaderContext.Provider value={{ isLoaderVisible, setLoaderVisibility }}>{children}</LoaderContext.Provider>;
};

export const useLoader = () => {
    const context = useContext(LoaderContext);

    if (context === undefined) {
        throw new Error('useLoader must be used within a LoaderProvider');
    }

    return context;
};
