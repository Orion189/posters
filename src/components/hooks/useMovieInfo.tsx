import { useState, ReactNode, FC, Dispatch, SetStateAction, useContext, createContext, useMemo } from 'react';
import { MovieInfo } from 'src/types';

type MovieInfoContextProps = {
    movieInfo: MovieInfo | undefined;
    setMovieInfo: Dispatch<SetStateAction<MovieInfo | undefined>>;
};

type MovieInfoProviderProps = {
    children?: ReactNode;
};

const MovieInfoContext = createContext<MovieInfoContextProps | undefined>(undefined);

export const MovieInfoProvider: FC<MovieInfoProviderProps> = ({ children }) => {
    const [movieInfo, setMovieInfo] = useState<MovieInfo | undefined>(undefined);
    const data = useMemo(() => ({ movieInfo, setMovieInfo }), [movieInfo]);

    return <MovieInfoContext.Provider value={data}>{children}</MovieInfoContext.Provider>;
};

export const useMovieInfo = () => {
    const context = useContext(MovieInfoContext);

    if (context === undefined) {
        throw new Error('useMovieInfo must be used within a CountProvider');
    }

    return context;
};
