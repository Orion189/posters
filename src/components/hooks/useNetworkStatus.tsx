import { useEffect, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useSnackbar } from 'notistack';

export const useNetworkStatus = () => {
    const { t } = useTranslation();
    const { enqueueSnackbar } = useSnackbar();
    const toggleOnline = useCallback(
        () => enqueueSnackbar(t('common.label.network.online'), { variant: 'success' }),
        []
    );
    const toggleOffline = useCallback(
        () => enqueueSnackbar(t('common.label.network.offline'), { variant: 'error' }),
        []
    );

    useEffect(() => {
        window.addEventListener('online', toggleOnline);
        window.addEventListener('offline', toggleOffline);
    }, []);
    useEffect(
        () => () => {
            window.removeEventListener('online', toggleOnline);
            window.removeEventListener('offline', toggleOffline);
        },
        []
    );
};
