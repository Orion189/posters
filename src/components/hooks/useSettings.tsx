import { useState, ReactNode, FC, Dispatch, SetStateAction, createContext, useContext } from 'react';

type SettingsContextProps = {
    isSettingsStored: boolean;
    setSettingsStored: Dispatch<SetStateAction<boolean>>;
};

type SettingsProviderProps = {
    children?: ReactNode;
};

const SettingsContext = createContext<SettingsContextProps | undefined>(undefined);

export const SettingsProvider: FC<SettingsProviderProps> = ({ children }) => {
    const [isSettingsStored, setSettingsStored] = useState(false);

    return (
        <SettingsContext.Provider value={{ isSettingsStored, setSettingsStored }}>{children}</SettingsContext.Provider>
    );
};

export const useSettings = () => {
    const context = useContext(SettingsContext);

    if (context === undefined) {
        throw new Error('useSettings must be used within a SettingsProvider');
    }

    return context;
};
