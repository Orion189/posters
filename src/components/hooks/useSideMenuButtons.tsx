import { useEffect, useRef, useCallback, MutableRefObject } from 'react';
import commonStyles from '@styles/common.module.scss';

export const useSideMenuButtons = (
    settingsBtnRef: MutableRefObject<HTMLButtonElement | null>,
    infoBtnRef: MutableRefObject<HTMLButtonElement | null>
) => {
    const timeoutRef = useRef<undefined | number>(undefined);
    const isBtnsShownRef = useRef<boolean>(false);
    const onMouseMove = useCallback((e: MouseEvent) => {
        const settingsBtn = settingsBtnRef?.current;
        const infoBtn = infoBtnRef?.current;
        let isBtnsShown = isBtnsShownRef?.current;
        let timeout = timeoutRef?.current;

        if (!isBtnsShown) {
            settingsBtn?.classList.add(commonStyles.fullOpacity);
            infoBtn?.classList.add(commonStyles.fullOpacity);
            isBtnsShown = true;
        }

        window.clearTimeout(timeout);
        timeout = window.setTimeout(() => {
            settingsBtn?.classList.remove(commonStyles.fullOpacity);
            infoBtn?.classList.remove(commonStyles.fullOpacity);
            isBtnsShown = false;
        }, 1000);
    }, []);

    useEffect(() => {
        window.addEventListener('mousemove', onMouseMove);
    }, [onMouseMove]);
    useEffect(
        () => () => {
            window.removeEventListener('mousemove', onMouseMove);
        },
        [onMouseMove]
    );
};
