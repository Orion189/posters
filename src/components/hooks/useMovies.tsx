import { useRef, useCallback, useEffect, useState } from 'react';
import getConfig from 'next/config';
import type { Movie } from 'src/types';
import { useSnackbar } from 'notistack';
import { useLazyQuery } from '@apollo/client';
import { useTranslation } from 'react-i18next';
import { useLoader } from '@components/hooks/useLoader';
import { useSettings } from '@components/hooks/useSettings';
import { useMovieInfo } from '@components/hooks/useMovieInfo';
import { GET_RANDOM_MOVIES, GET_MOVIES_BY_SETTINGS } from 'src/graphql/queries';

const TIMEOUT = 1000 * 30;
const { publicRuntimeConfig } = getConfig();

export const useMovies = () => {
    const { t } = useTranslation();
    const { enqueueSnackbar } = useSnackbar();
    const { isSettingsStored } = useSettings();
    const { setMovieInfo } = useMovieInfo();
    const { setLoaderVisibility } = useLoader();
    const timeoutRef = useRef<number | undefined>(undefined);
    const [movies, setMovies] = useState<Movie[]>([]);
    const [index, setIndex] = useState<number | undefined>(undefined);
    const [getRandomMovies] = useLazyQuery(GET_RANDOM_MOVIES, {
        fetchPolicy: 'network-only',
        nextFetchPolicy: 'network-only',
        onCompleted: (data) => {
            if (Array.isArray(data?.random) && data?.random.length > 0) {
                setMovies(data?.random);
                setIndex(0);
                setMovieInfo(movies[0]);
            }
        }
    });
    const [getMoviesBySettings] = useLazyQuery(GET_MOVIES_BY_SETTINGS, {
        fetchPolicy: 'network-only',
        nextFetchPolicy: 'network-only',
        onCompleted: (data) => {
            if (Array.isArray(data?.search) && data?.search.length > 0) {
                setMovies(data.search);
                setIndex(0);
                setMovieInfo(movies[0]);
            }
        }
    });

    const changeMovie = useCallback(async () => {
        const isMovieAvailable = index !== undefined && movies[index + 1];

        if (isMovieAvailable) {
            setIndex(index + 1);
            setMovieInfo(movies[index + 1]);
        } else {
            const settingsStr = window.localStorage.getItem(publicRuntimeConfig.lsItemName);

            if (settingsStr) {
                try {
                    getMoviesBySettings({
                        variables: JSON.parse(settingsStr)
                    });
                } catch (e) {
                    enqueueSnackbar(t('common.errors.settings.parsing'), { variant: 'error' });
                }
            } else {
                getRandomMovies();
            }
        }

        setLoaderVisibility(false);
    }, [index]);

    useEffect(() => {
        setMovies([]);
        setLoaderVisibility(true);
    }, [isSettingsStored]);

    useEffect(() => {
        window.clearTimeout(timeoutRef.current);
        timeoutRef.current = window.setTimeout(changeMovie, TIMEOUT);
    }, [index]);

    useEffect(() => {
        setLoaderVisibility(true);
    }, []);

    useEffect(
        () => () => {
            window.clearTimeout(timeoutRef.current);
        },
        []
    );

    return {
        index,
        movies
    };
};
