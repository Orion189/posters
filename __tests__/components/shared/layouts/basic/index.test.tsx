import { cleanup, render, act, RenderResult } from '@testing-library/react';
import { LoaderProvider } from '@components/hooks/useLoader';
import BasicLayout from '@components/shared/layouts/basic';

let useNetworkStatus: jest.Mock;
let useSideMenuButtons: jest.Mock;

jest.mock('@components/hooks/useNetworkStatus', () => ({
    useNetworkStatus: () => useNetworkStatus()
}));
jest.mock('@components/hooks/useSideMenuButtons', () => ({
    useSideMenuButtons: () => useSideMenuButtons()
}));

describe('BasicLayout', () => {
    let result: RenderResult;

    beforeEach(() => {
        useNetworkStatus = jest.fn();
        useSideMenuButtons = jest.fn();

        act(() => {
            result = render(
                <LoaderProvider>
                    <BasicLayout>
                        <div data-testid="test-element"></div>
                    </BasicLayout>
                </LoaderProvider>
            );
        });
    });

    afterEach(() => {
        jest.restoreAllMocks();
        cleanup();
    });

    it('Should render children components', () => {
        const { getByTestId } = result;
        const testElement = getByTestId('test-element');

        expect(testElement).toBeInTheDocument();
    });
    
    it('Should call `useNetworkStatus` and `useSideMenuButtons` hooks', () => {
        expect(useNetworkStatus).toHaveBeenCalledTimes(1);
        expect(useSideMenuButtons).toHaveBeenCalledTimes(1);
    });

    it('Should render <BasicLayout /> unchanged', () => {
        const { container } = result;

        expect(container).toMatchSnapshot()
    });
});
