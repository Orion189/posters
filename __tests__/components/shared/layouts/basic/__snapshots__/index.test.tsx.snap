// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicLayout Should render <BasicLayout /> unchanged 1`] = `
<div>
  <div
    class="wrapper MuiBox-root css-0"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-1q8n7u2-MuiBackdrop-root"
      style="opacity: 0; visibility: hidden;"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1g9z2ko-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
    <button
      aria-label="settings"
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary settingsBtn css-1p0gq2-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-1ezohtm-MuiSvgIcon-root"
        data-testid="SettingsRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.5 12c0-.23-.01-.45-.03-.68l1.86-1.41c.4-.3.51-.86.26-1.3l-1.87-3.23c-.25-.44-.79-.62-1.25-.42l-2.15.91c-.37-.26-.76-.49-1.17-.68l-.29-2.31c-.06-.5-.49-.88-.99-.88h-3.73c-.51 0-.94.38-1 .88l-.29 2.31c-.41.19-.8.42-1.17.68l-2.15-.91c-.46-.2-1-.02-1.25.42L2.41 8.62c-.25.44-.14.99.26 1.3l1.86 1.41c-.02.22-.03.44-.03.67s.01.45.03.68l-1.86 1.41c-.4.3-.51.86-.26 1.3l1.87 3.23c.25.44.79.62 1.25.42l2.15-.91c.37.26.76.49 1.17.68l.29 2.31c.06.5.49.88.99.88h3.73c.5 0 .93-.38.99-.88l.29-2.31c.41-.19.8-.42 1.17-.68l2.15.91c.46.2 1 .02 1.25-.42l1.87-3.23c.25-.44.14-.99-.26-1.3l-1.86-1.41c.03-.23.04-.45.04-.68zm-7.46 3.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      data-testid="test-element"
    />
    <button
      aria-label="info"
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary infoBtn css-1p0gq2-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
        data-testid="InfoRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 15c-.55 0-1-.45-1-1v-4c0-.55.45-1 1-1s1 .45 1 1v4c0 .55-.45 1-1 1zm1-8h-2V7h2v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
